File: backend/app/api/endpoints/file.py
import os
import json
import datetime
from fastapi import APIRouter
from fastapi.responses import FileResponse

# 固定保存目录
RESULT_DIR = "/home/lys/projects/automated_design/backend/app/utils/analysis/result"
os.makedirs(RESULT_DIR, exist_ok=True)

router = APIRouter()


def write_txt(filename: str, data: dict):
    """写入 txt 文件"""
    filepath = os.path.join(RESULT_DIR, filename)
    with open(filepath, "w", encoding="utf-8") as f:
        f.write(json.dumps(data, indent=2, ensure_ascii=False))
    return filepath


@router.post("/save/block_matrix/")
async def save_block_matrix(data: dict):
    filename = f"block_matrix_{datetime.datetime.now().strftime('%Y%m%d%H%M%S')}.txt"
    filepath = write_txt(filename, data)
    return {"code": 0, "msg": "分块阵保存成功", "filename": filename, "path": filepath}


@router.post("/save/bit_matrix/")
async def save_bit_matrix(data: dict):
    filename = f"bit_matrix_{datetime.datetime.now().strftime('%Y%m%d%H%M%S')}.txt"
    filepath = write_txt(filename, data)
    return {"code": 0, "msg": "比特阵保存成功", "filename": filename, "path": filepath}


@router.post("/save/sbox/")
async def save_sbox(data: dict):
    filename = f"sbox_{datetime.datetime.now().strftime('%Y%m%d%H%M%S')}.txt"
    filepath = write_txt(filename, data)
    return {"code": 0, "msg": "S盒保存成功", "filename": filename, "path": filepath}


@router.get("/download/{filename}")
async def download(filename: str):
    filepath = os.path.join(RESULT_DIR, filename)
    if os.path.exists(filepath):
        return FileResponse(filepath, filename=filename, media_type="text/plain")
    return {"code": 1, "msg": "文件不存在"}

File: frontend/src/apis/file.ts
import { request, ApiResponse } from './request'

// 和后端保持一致
export interface SaveResponse {
  code: number
  msg: string
  filename: string
  path: string
}

// =================== 保存 ===================
export const saveBlockMatrixAPI = async (data: any): Promise<SaveResponse> => {
  const res: ApiResponse<SaveResponse> = await request<SaveResponse>({
    url: '/file/save/block_matrix/',
    method: 'POST',
    data,
  })
  // 关键：从 ApiResponse<T> 解包成 SaveResponse
  return res as unknown as SaveResponse
}

export const saveBitMatrixAPI = async (data: any): Promise<SaveResponse> => {
  const res: ApiResponse<SaveResponse> = await request<SaveResponse>({
    url: '/file/save/bit_matrix/',
    method: 'POST',
    data,
  })
  return res as unknown as SaveResponse
}

export const saveSboxAPI = async (data: any): Promise<SaveResponse> => {
  const res: ApiResponse<SaveResponse> = await request<SaveResponse>({
    url: '/file/save/sbox/',
    method: 'POST',
    data,
  })
  return res as unknown as SaveResponse
}

// =================== 下载 ===================
export const downloadFileAPI = async (filename: string): Promise<Blob> => {
  const res = await request<any>({
    url: `/file/download/${filename}`,
    method: 'GET',
    responseType: 'blob',
  })
  return res as unknown as Blob
}

